src/client/datascience/notebook/vscodeNotebookController.ts

```diff
@@ -237,6 +237,11 @@ export class VSCodeNotebookController implements Disposable {
             // Possible user selected a controller that's not contributed by us at all.
             const kernel = this.kernelProvider.get(event.notebook);
             if (kernel?.kernelConnectionMetadata.id === this.kernelConnection.id) {
+                traceInfo(
+                    `Disposing kernel ${this.kernelConnection.id} for notebook ${getDisplayPath(
+                        event.notebook.uri
+                    )} due to selection of another kernel or closing of the notebook`
+                );
```

Comment:
This was crucial, was difficult to tell why a kernel was disposed.
@rchiodo @IanMatthewHuff FYI - When closing a notebook, VSC un-selects a controller, hence kernels get disposed here.
I wasn't expecting that at all, I was hoping the notebook closed event would get triggered first, but thats not true.